### PR TITLE
Backport all Painless compiler extensibility changes

### PR DESCRIPTION
**Pull request by jdconrad**
This backport incorporates all the changes to improve compiler extensibility. The reason for this backport is the changes are now required to support runtime fields.

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/oss-distro-docs

File: modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java [hunk 1]
```diff
@@ -460,11 +460,11 @@ public static Whitelist loadFromResourceFiles(Class<?> resource, Map<String, Whi
 
         List<Object> annotations;
 
-        line = line.trim();
-
-        if (line.isEmpty()) {
+        if ("".equals(line.replaceAll("\\s+",""))) {
```
Comment (stu-elastic) [inline modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java:463]: This is `line.isBlank()` in master, should master be updated?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java [hunk 2]
```diff
@@ -284,15 +284,8 @@ static Request parse(XContentParser parser) throws IOException {
         Request(StreamInput in) throws IOException {
             super(in);
             script = new Script(in);
-            if (in.getVersion().before(Version.V_6_4_0)) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:287]: Do we need to keep the 6.4 back compat code here?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java [hunk 3]
```diff
@@ -328,12 +321,8 @@ public ActionRequestValidationException validate() {
         public void writeTo(StreamOutput out) throws IOException {
             super.writeTo(out);
             script.writeTo(out);
-            if (out.getVersion().before(Version.V_6_4_0)) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:331]: Do we need to keep the 6.4 back compat code here?

File: modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java:463]: isBlank isn't available in 8

Comment (stu-elastic) [reply modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java:463]: No, doesn't exist in 8.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:287]: Good catch. Will change back.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java:331]: Same as above.

File: modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java [hunk 4]
```diff
@@ -502,20 +502,20 @@ public static Whitelist loadFromResourceFiles(Class<?> resource, Map<String, Whi
                         if (argumentKeyValue.length != 2) {
                             throw new IllegalArgumentException("invalid annotation: expected key=\"value\" [" + line + "]");
                         }
-                        
+
```
Comment (stu-elastic) [inline modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java:505]: After we get this into 7.x, let's fix these up in master to cleanup the backport.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/110_script_score_boost.yml [hunk 5]
```diff
@@ -1,8 +1,8 @@
 # Integration tests for ScriptScoreQuery using Painless
 setup:
   - skip:
-      version: " - 7.6.99"
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/110_script_score_boost.yml:4]: Take a pass on these yml changes.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/15_update.yml [hunk 6]
```diff
@@ -20,7 +20,6 @@
               params: { bar: 'xxx' }
 
   - match: { _index:   test_1 }
-  - match: { _type:    _doc   }
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/15_update.yml:23]: Take a pass on these yml changes.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/20_scriptfield.yml [hunk 7]
```diff
@@ -151,6 +151,6 @@ setup:
     - match: { error.type: "search_phase_execution_exception" }
     - match: { error.reason: "all shards failed" }
     - match: { error.failed_shards.0.reason.caused_by.type: "illegal_argument_exception" }
-    - match: { error.failed_shards.0.reason.caused_by.reason: "While loop has no escape." }
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/20_scriptfield.yml:154]: Take a pass on these yml changes.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml [hunk 8]
```diff
@@ -452,10 +452,6 @@
 ---
 
 "Exception on negative score":
-    - skip:
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml:455]: Take a pass on these yml changes.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/70_mov_fn_agg.yml [hunk 9]
```diff
@@ -1,9 +1,6 @@
 # Sanity integration test to make sure the custom context and whitelist work for moving_fn pipeline agg
 #
 setup:
-  - skip:
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/70_mov_fn_agg.yml:4]: Take a pass on these yml changes.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/80_script_score.yml [hunk 10]
```diff
@@ -1,10 +1,5 @@
 # Integration tests for ScriptScoreQuery using Painless
 
-setup:
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/80_script_score.yml:3]: Take a pass on these yml changes.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/90_interval_query_filter.yml [hunk 11]
```diff
@@ -1,8 +1,4 @@
 setup:
-  - skip:
```
Comment (stu-elastic) [inline modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/90_interval_query_filter.yml:2]: Take a pass on these yml changes.

**Review by stu-elastic — APPROVED** [review]
Looks good compared to master, all changes from master are Java 8 compat issues.

However, it looks like we removed some stuff that was unintentional, such as wire compat and yml tests.

File: modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/spi/src/main/java/org/elasticsearch/painless/spi/WhitelistLoader.java:505]: Agreed, we should just do an entire whitespace PR.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/15_update.yml (hunk 6 above)
Comment (jdconrad) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/15_update.yml:23]: Reverted.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/20_scriptfield.yml (hunk 7 above)
Comment (jdconrad) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/20_scriptfield.yml:154]: Required as the error message changed.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml (hunk 8 above)
Comment (jdconrad) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/30_search.yml]: Reverted.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/70_mov_fn_agg.yml (hunk 9 above)
Comment (jdconrad) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/70_mov_fn_agg.yml:4]: Reverted.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/80_script_score.yml (hunk 10 above)
Comment (jdconrad) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/80_script_score.yml:3]: Reverted.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/90_interval_query_filter.yml (hunk 11 above)
Comment (jdconrad) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/90_interval_query_filter.yml]: Reverted.

File: modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/110_script_score_boost.yml (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/110_script_score_boost.yml:4]: Reverted.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for reviewing this. Will commit as soon as CI passes.

**Pull request merged**